cpp/src/gandiva/like_holder.cc

```diff
@@ -62,39 +56,95 @@ const FunctionNode LikeHolder::TryOptimize(const FunctionNode& node) {
   return node;
 }
 
+const FunctionNode SQLLikeHolder::TryOptimize(const FunctionNode& node) {
```

Comment:
`like` will now always optimize into `regexp_matches` or one of the optimizations above. I found this to be the easiest way to reuse the optimization code.